### ISSUE TITLE
Enable a form of -Xshareclasses by default

### ISSUE THREAD

**Issue by pshipton** [eclipse/openj9#1646]
A form of -Xshareclasses should be enabled by default in order to provide startup time and footprint advantages.

There are a number of factors to consider:

1. what classloaders should be shared by default? -Xshareclasses enables URLClassLoader *and potentially subclasses of URLClassLoader* for class sharing. As we saw in #1244, the subclasses don't necessarily work, resulting in the JVM failing to start. Restricting to the bootstrap loader would resolve this, as would restricting to the bootstrap loader plus the application and extension (or platform) loaders. Given the application code and extension (or platform) code may be different from multiple JVM invocations, sharing only the bootstrap loader would be the safest option. 

1. what should the cache be called? The cache will be shared by every invocation of the JVM from the same user. The username should be part of the cache name, which it already is by default. The shared cache is automatically destroyed (when possible) and recreated when the "unique id" of the VM attaching to the cache is different from the "unique id" of the VM which created the cache. Using the "unique id" as the cache name will ensure that only identical builds of the JVM can shared the cache. (fyi #1645).

1. what happens to abandoned default caches? Abandoned either because the JVM was updated, or the user added an explicit -Xshareclasses option. They will consume disk space, or on z/OS shared memory. There is an `-Xshareclasses:expire=<time in minutes>` option which destroys caches that haven't been used in the number of minutes specified. The VM could automatically destroy any default caches it finds which are unused for some period of time.

**Comment by pshipton** [comment]
Yes, AOT will be enabled with the shared cache. As it stands, cache files are stored in /tmp/javasharedresources and have the user name (j9build in this case) as part of the cache name. i.e. /tmp/javasharedresources/C290M5F1A64P_sharedcc_j9build_G35

**Comment by ashu-mehra** [comment]
I kind of like James suggestion to store file under $HOME which is similar to the way other software store their config files.

**Comment by hangshao0** [comment]
It might be possible to allow JVM to attach to 2 shared caches, one for Bootstrap Classloader and one for other class loaders. Currently the whole shared cache will be deleted if a corruption is detected, we will lose all the shared classes in it.

**Comment by DanHeidinga** [comment]
How complex does that make it for the end user?  Especially when trying to figure out what's going wrong.

**Comment by hangshao0** [comment]
> How complex does that make it for the end user?

You are asking about enable -Xshareclasses by default or allow JVM to attach to 2 shared caches ?

**Comment by DanHeidinga** [comment]
Attaching to 2 shared caches.

**Comment by DanHeidinga** [comment]
We should walk through some of the user scenarios for diagnosing a problem with the boot vs non-boot cache, for naming one or both, for sizing them, soft-limit, etc.  One way to start would be to look at all the SCC options and see how they would have to be adapted to work in a multi-cache world.

**Comment by hangshao0** [comment]
Are we doing this (enabling -Xshareclasses by default) on Java 9+ or Java 8+ ? OpenJ9 Only or OpenJ9 + IBM SDK ?

**Comment by DanHeidinga** [comment]
Let's start by enabling by default for Java 9+.  

> or OpenJ9 + IBM SDK ?

That's a question for IBM to decide what they want to do with their SDK.

**Comment by hangshao0** [comment]
I will first enabling the bootstrap class sharing by default (which is WIP). Then looking at the multi-SCC scenario. In the multi-SCC scenario, I guess user does not need to tune the default bootstrap classes SCC and the the current -Xshareclasses option can be applied to the non-bootstrap classes SCC. Once I have more info, I will post them here.

**Comment by hangshao0** [comment]
Also since Java 9 is obsolete, the bootstrap class sharing will be enabled in Java 10+ in my WIP pull request.

**Comment by vijaysun-omr** [comment]
@mstoodle had an interesting question about this feature : should we consider enabling it in Java8 specifically in the OpenJ9 release (that is still a relatively new offering) ? 

I suppose a similar question could be asked wrt changing the shared classes cache default size that was done some time back.

**Comment by hangshao0** [comment]
I am fine with enabling bootstrap class sharing by default on OpenJ9 Java 8. Probably we can enable it on Java 10+ first and see how it goes. If everything is fine, we can backport onto OpenJ9 Java 8.

There is an open issue (#1938) about increasing the default shared cache size on Java 8.

**Comment by ashu-mehra** [comment]
fyi - openjdk will soon have default CDS archive in th JDK as per https://bugs.openjdk.java.net/browse/JDK-8204247.
Does this issue also aim at providing similar feature, i.e. having a default shared class cache as part of the builds?

**Comment by hangshao0** [comment]
> Does this issue also aim at providing similar feature, i.e. having a default shared class cache as part of the builds?

I don't see such request in the description of this issue. 
If we want to do something similar, we can pre-populate a shared cache with all the bootstrap classes and put it into JDK. All the bootstrap classes takes about 40MB (Java 8).

**Comment by DanHeidinga** [comment]
Let's keep this issue for enabling SCC by default.  We can discuss prepopulating the cache at build time in a separate issue.

For JDK11+, my preference would be to take another look at jlink and jimage to determine how to better take advantage of those tools.

**Comment by mpirvu** [comment]
We will need performance evaluation of benchmarks that don't use Xshareclasses by default (anything that doesn't use Liberty and WAS). This is because AOT is going to be enabled by default and this triggers JIT changes that may favor start-up time over peak throughput. For instance, the default method invocation counts will change from 3000,3000 to 1000,250 resulting in fewer interpreter profiler samples which is turn has the potential of affecting throughput.

**Comment by hangshao0** [comment]
> is this feature something that is targetting the OpenJ9 Release 0.10.0 milestone ?

I don't see the 0.10.0 milestone in this issue.

> Please work with Marius to get the evaluations he mentioned in the previous comment done (you can point him at a build and he can help).

Yes, I will point Marius to the build once is ready.

**Comment by ashu-mehra** [comment]
> One question that was raised in a discussion about this feature was the behaviour in Docker. The concern specifically was around the value of creating such an SCC by default in such an environment where we may not get a chance to reuse the SCC.

I think this can be addressed by checking if the JVM is running in a container or not. If in container, we can resort to current behavior.

**Comment by DanHeidinga** [comment]
Assigning this to milestone 0.11.0 based on discussions with @hangshao0

**Comment by hangshao0** [comment]
> Please work with Marius to get the evaluations he mentioned in the previous comment done (you can point him at a build and he can help).

@mpirvu , I have created a build and shared that to you.

**Comment by SueChaplain** [comment]
I assume this work will require doc?

**Comment by pshipton** [comment]
Yup, done.

**Comment by hangshao0** [comment]
> Please also create an issue at the openj9-docs repo

I will open an issue at the openj9-docs repo once we have decided on the details and finalized the implementation.

**Comment by hangshao0** [comment]
@mpirvu , Do you think there is a need to set a maxAOT limit for the default shared cache ?

**Comment by mpirvu** [comment]
No, I think we should not set any limit on AOT.

**Comment by hangshao0** [comment]
> No, I think we should not set any limit on AOT.

I have built an OpenJ9 Java 8 SDK with -Xshareclasses:bootClassesOnly and tried on Eclipse. I see very time I startup Eclipse, the number of the romclasses in the shared cache remain the same after the cold run. But there are about 10 to 20 more methods being AOT'ed and stored into the shared cache every time after the cold run. This means the default cache could be gradually filled with AOT. Is this a problems ? @mpirvu

**Comment by mpirvu** [comment]
Does that AOT number slowly decrease over time? 
It is possible that some methods that haven't been compiled in the first run will be compiled in the additional runs, but that trend should gradually go down to 0. 

Note that in some liberty workloads I have seen new classes being added to the SCC with additional run which I find weird.

**Comment by pshipton** [comment]
It might help to list the methods being AOTed for the first 5 or 10 iterations.

**Comment by hangshao0** [comment]
[AOT Stats.txt](https://github.com/eclipse/openj9/files/2425902/AOT.Stats.txt)

**Comment by hangshao0** [comment]
AOT methods are listed in AOT stats.txt.

**Comment by mpirvu** [comment]
Actually what is more interesting is the set of methods that get AOTed in iteration 5, 6, 7 etc. Then we could look for patterns.

**Comment by pshipton** [comment]
They are there in the attached file. The number at the start of each line is the iteration number. Do you need more iterations?

**Comment by mpirvu** [comment]
Oh, I didn't realize the meaning of that number.
In the file I see I see 1,2,3,6,8. I wonder if the absence of 4,5,7 is meaningful in some way.

I cannot find any patterns from that file. Methods compiled late belong to different classes including java/lang and java/util. If the runs were not executed with -Xaot:forceaot (as I guess) then the JIT heuristics that decide what to AOT may play a role here because timing is slightly different from run to run.
A while back the JIT had the policy of AOTing only in the cold run. This has the disadvantage that if two JVM executing different applications connect to the same SCC, the second JVM/application may not get the chance to AOT its specific methods.

A more fundamental question is whether this seemingly perpetual AOTing is a bad thing. I would argue that as long as the AOTed methods are used this is not necessarily bad. Thus, I propose the following experiment:
1) Run 50 iterations and at the 50th look at the JIT stats showing how many AOT methods were in the SCC and how many were loaded from SCC.
2) Without destroying the SCC run another 50 iterations and again look at JIT stats. If we see that the number of AOTed methods increased substantially, but the number of loaded AOT bodies stayed the same this could be an argument to bring back the policy for AOTing only in the cold run.

The way to enable JIT stats is to define the following environment variable
`export TR_PrintCompStats=1`
The stats will be printed to stderr at JVM shutdown (thus the JVM must be shutdown cleanly and not killed).

**Comment by hangshao0** [comment]
50th run:
```
Number of compilations per level:
Level=1 numComp=885
Level=2 numComp=313
Level=3 numComp=1
NumJNIMethodsCompiled=130
NumMethodsFoundInSharedCache=2300
NumMethodsTakenFromSharedCache=2268
NumAotedMethods=6
NumberOfAotedMethodsThatWereRecompiled=44 (forced=44)
```
100th run:
```
Number of compilations per level:
Level=1 numComp=885
Level=2 numComp=340
Level=3 numComp=1
NumJNIMethodsCompiled=129
NumMethodsFoundInSharedCache=2524
NumMethodsTakenFromSharedCache=2481
NumAotedMethods=2
NumberOfAotedMethodsThatWereRecompiled=60 (forced=60)
```
Both NumMethodsFoundInSharedCache and NumMethodsTakenFromSharedCache gradually increase. The only Java options I'm using is -Xshareclasses:name=BootClasses,BootClassesOnly,CacheDir=dir -Xscmx64m.

**Comment by mpirvu** [comment]
So the number of AOTed methods stored increased by 224 and the number of methods loaded increased by 213. This is good: the additional methods are actually used and not wasted, so startup time could improve over time. I do not think this is a problem.

**Comment by pshipton** [comment]
One concern is that the shared cache default size is 64MB, and it may gradually fill up over time to consume this entire space on disk.

@hangshao0 another concern I don't think we raised earlier, there is only a single cache used even if multiple JVM builds (different shas) are being used. Every time a JVM with a different sha starts, it will destroy the cache and recreate it.

**Comment by pshipton** [comment]
Reopening since #3295 has been reverted and likely #2449 will be as well.

**Comment by pshipton** [comment]
Moving to the 0.12.0 milestone since its too late to fix this in 0.11.0

**Comment by pshipton** [comment]
The next revision of this should enable the Platform ClassLoader by default as well.
